### PR TITLE
Use Version Metadata in buildagainstpackages if specified and remove AssemblyVersion

### PR DESCRIPTION
**Pull request by karajas**
Ensure version returned contains prerelease label.

### CODE REVIEW

**Comment by karajas** [comment]
@weshaggard PTAL.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 1]
```diff
@@ -358,14 +358,19 @@ private JObject GenerateDependencies(JObject projectJsonRoot, ITaskItem[] extern
                 {
                     NuGetVersion nuGetVersion = NuGetVersion.Parse(dependency.GetMetadata("Version"));
                     PackageItem packageItem = new PackageItem(name, nuGetVersion);
+                    bool? forceVersion = dependency.GetMetadata("ForceVersion")?.Equals("true", StringComparison.OrdinalIgnoreCase);
                     string version = packageItem.GetVersionString();
 
                     // a package version was provided, use its version information.
-                    if (packageInformation.ContainsKey(name))
+                    if (packageInformation.ContainsKey(name) && forceVersion!=true)
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I think I would prefer `!forceVersion` as it is a little clearer.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 2]
```diff
@@ -358,14 +358,19 @@ private JObject GenerateDependencies(JObject projectJsonRoot, ITaskItem[] extern
                 {
                     NuGetVersion nuGetVersion = NuGetVersion.Parse(dependency.GetMetadata("Version"));
                     PackageItem packageItem = new PackageItem(name, nuGetVersion);
+                    bool? forceVersion = dependency.GetMetadata("ForceVersion")?.Equals("true", StringComparison.OrdinalIgnoreCase);
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I'm not sure I fully understand what this is going to do. It isn't clicking in my head where ForceVersion will show-up and how it actually applies to the rewritting. @chcosta should also look at this change.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 3]
```diff
@@ -458,7 +463,7 @@ public NuGetVersion Version
 
         public string GetVersionString()
         {
-            return string.Join(".", _version.Major, _version.Minor, _version.Patch);
```
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: Do you no longer require a 3-part version string?  older versions of NuGet used a 4 part version string.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs (hunk 2 above)
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: What are you forcing the version to be and why?  ie, why would you not want to use the version provided by a package drop or version file and would instead prefer the assembly version?

Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I see, this is related to https://github.com/dotnet/corefx/pull/13128/files...

Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I wonder if it makes more sense to remove https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets#L132 and instead always use "Version" if it is present.

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: This was part of the workaround to get all the runtime dependencies with the right version (LKG). I'm ok with using Version if present if that's cleaner.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs (hunk 3 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I wanted to include the prerelease label. Will do some more testing to make sure we don't break.

**Comment by karajas** [comment]
@chcosta I wanted an explicit way to test against packages that aren't built locally and require the use of an LKG. Currently the generated project.json will always consume the live version of a package with all live subdependencies. This makes sense in the context of build pipeline since we build each leg and publish and then consume these changes. Locally we want to use the LKG version. This seemed like the cheapest way around this issue for the generated proejct.json.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs (hunk 2 above)
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I'd prefer to use Version if it's present.

**Comment by weshaggard** [comment]
I believe with our current plans we can close this PR now.

**Comment by karajas** [comment]
@weshaggard @chcosta Updated this to have the regex fixes

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 4]
```diff
@@ -374,15 +374,21 @@ private JObject GenerateDependencies(JObject projectJsonRoot, ITaskItem[] extern
...
                 if (!returnDependenciesList.ContainsKey(name))
                 {
-                    NuGetVersion nuGetVersion = NuGetVersion.Parse(dependency.GetMetadata("Version"));
-                    PackageItem packageItem = new PackageItem(name, nuGetVersion);
-                    string version = packageItem.GetVersionString();
+                    string versionFromMetadata = dependency.GetMetadata("Version");
+                    NuGetVersion nuGetVersion = null;
+                    if (versionFromMetadata != string.Empty)
+                    {
+                        nuGetVersion = NuGetVersion.Parse(dependency.GetMetadata("Version"));
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: dependency.GetMetadata("Version") -> versionFromMetadata

**Comment by karajas** [comment]
Abandoning this for now. Will cleanup later.

**Pull request closed without merging**